Csrf and hpos

to6ko12 opened the pull request:
[sc-28984]

to6ko12 committed:
Add hpos compatibility

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -60,8 +60,11 @@ public function autoSyncNotice() {
+            /** @var WC_Order $order */
+            $order = wc_get_order($order_id);
+        if (!$order->get_meta('wpml_language') && defined('ICL_LANGUAGE_CODE')) {
+            $order->get_meta_data();
```
Is there a reason for ` $order->get_meta_data();`

TsankoBg commented on webwinkelkeur/webwinkelkeur.php:
```diff
@@ -19,5 +19,11 @@
     return;
 }
 
+add_action( 'before_woocommerce_init', function() {
+    if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
+        \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
+    }
+} );
+
```
Add this to `BasePlugin.php`

TsankoBg reviewed (changes requested):
As discussed in telegram, could you do very extensive testing and research about the support of the used woocommerce functions?

to6ko12 committed:
Cleanup + add checks
1. move compatibility check to right file
2. add checks if hpos is enabled and act accordingly

to6ko12 committed:
Check if class exists
To ensure backwards compatibility

to6ko12 committed:
Cleanup

TsankoBg commented:
@to6ko12 create a package and provide it to CS to test it after it it's been tested, we will publish the update

apeschar commented on common/src/BasePlugin.php:
```diff
@@ -36,6 +36,13 @@ public function init() {
         add_action('plugins_loaded', [$this, 'loadTranslations']);
         add_action('admin_enqueue_scripts', [$this, 'addUpdateNoticeDismissScript']);
         add_action('wp_ajax_' . $this->getUpdateNoticeDismissedAjaxHook(), [$this, 'dismissUpdateNotice']);
+        add_action( 'before_woocommerce_init', function() {
+            if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
+                \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility(
+                    'custom_order_tables', $this->getSlug() . '/' . $this->getSlug() . '.php', true
+                );
+            }
+        } );
```
Please don't apply WordPress coding standards. That is, drop all the extra spaces.

apeschar commented on common/src/BasePlugin.php:
```diff
@@ -36,6 +36,13 @@ public function init() {
         add_action('admin_enqueue_scripts', [$this, 'addUpdateNoticeDismissScript']);
         add_action('wp_ajax_' . $this->getUpdateNoticeDismissedAjaxHook(), [$this, 'dismissUpdateNotice']);
+        add_action( 'before_woocommerce_init', function() {
+            if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
+                \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility(
```
Better add a `use` statement for this long class name.

apeschar commented on common/src/WooCommerce.php:
```diff
@@ -60,8 +61,16 @@ public function autoSyncNotice() {
-        if (!get_post_meta($order_id, 'wpml_language') && defined('ICL_LANGUAGE_CODE')) {
-            update_post_meta($order_id, 'wpml_language', ICL_LANGUAGE_CODE);
+        if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
+            /** @var WC_Order $order */
+            $order = wc_get_order($order_id);
+            if (!$order->get_meta('wpml_language') && defined('ICL_LANGUAGE_CODE')) {
+                $order->update_meta_data('wpml_language', ICL_LANGUAGE_CODE);
+            }
```
This technique should always work, independently of whether the custom orders table is used. So I think that the `get_post_meta`/`update_post_meta` code can be dropped.

to6ko12 replied:
As far as I read here:
https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book
these should be changed, or have I misunderstood it?

apeschar replied:
My point is that we don't need both. We only need the new API, the old API calls can be dropped.

In other words, I believe we can use the new API calls instead of the old API calls, not in addition to.

TsankoBg replied:
I believe @apeschar meant that you can use `get_meta` and `update_meta_data` without checking if `custom_orders_table_usage_is_enabled`

---

Also in the https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book
It says that `update_meta_data` should be followed by `order->save()`

apeschar commented on common/src/WooCommerce.php:
```diff
@@ -60,8 +61,16 @@ public function autoSyncNotice() {
     public function set_order_language($order_id) {
-        if (!get_post_meta($order_id, 'wpml_language') && defined('ICL_LANGUAGE_CODE')) {
-            update_post_meta($order_id, 'wpml_language', ICL_LANGUAGE_CODE);
+        if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```
Please remove the extra WordPress spaces.

apeschar commented on common/src/WooCommerce.php:
```diff
@@ -111,8 +120,7 @@ private function sendInvite($order_id) {
         ];
-
-        $lang = get_post_meta($order_id, 'wpml_language', true);
+        $lang = class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ? $order->get_meta('wpml_language', true) : get_post_meta($order_id, 'wpml_language', true);
```
`$order->get_meta()` is always fine to use, you don't need `get_post_meta`.

apeschar reviewed (changes requested):
This should be rebased on `master`.

to6ko12 committed:
Stopped using WP coding standards + remove unnecessary old API calls

to6ko12 committed:
Remove unnecessary check + save order if meta data is updated

apeschar commented on common/src/BasePlugin.php:
```diff
@@ -2,6 +2,7 @@
 namespace Valued\WordPress;
 
 use ReflectionClass;
+use \Automattic\WooCommerce\Utilities\FeaturesUtil;
```
`use` declarations don't need a leading `\`.

```suggestion
use Automattic\WooCommerce\Utilities\FeaturesUtil;
```

apeschar commented on common/src/BasePlugin.php:
```diff
@@ -36,6 +37,13 @@ public function init() {
+        add_action('before_woocommerce_init', function() {
+            if (class_exists(FeaturesUtil::class)) {
+                FeaturesUtil::declare_compatibility(
+                    'custom_order_tables', $this->getSlug() . '/' . $this->getSlug() . '.php', true
```
`getPluginFile` will return the same thing, and also handle the situation where the plugin directory is renamed. This is possible in WordPress, and it should still work in that case.

```suggestion
                    'custom_order_tables', $this->getPluginFile(), true
```

apeschar commented on common/src/WooCommerce.php:
```diff
@@ -60,8 +61,11 @@ public function autoSyncNotice() {
+        $order = wc_get_order($order_id);
+        if (!$order->get_meta('wpml_language') && defined('ICL_LANGUAGE_CODE')) {
+            $order->update_meta_data('wpml_language', ICL_LANGUAGE_CODE);
+            $order->save();
```
I had a look at the WooCommerce source code and it looks like `save_meta_data()` does a bit less than `save()` and prevents some caches from being cleared. So I suggest we use that instead.

```suggestion
            $order->save_meta_data();
```

apeschar replied:
This is marked as resolved, but I still see `save()` instead of `save_meta_data()`?

apeschar commented on common/src/WooCommerce.php:
```diff
@@ -8,6 +8,7 @@
 use WC_Product_Factory;
 use WP_Comment_Query;
 use WC_Comments;
+use Automattic\WooCommerce\Utilities\OrderUtil;
```
This `use` declaration is not actually necessary-- OrderUtil isn't used.

to6ko12 committed:
Code cleanup

to6ko12 committed:
Change save method of order meta data

apeschar commented on common/src/WooCommerce.php:
```diff
@@ -111,8 +114,7 @@ private function sendInvite($order_id) {
         ];
-
-        $lang = get_post_meta($order_id, 'wpml_language', true);
+        $lang = $order->get_meta('wpml_language', true);
```
Looks like the second parameter (`$single`) is true by default, so you don't need to specify it.

apeschar commented on common/src/BasePlugin.php:
```diff
@@ -36,6 +37,13 @@ public function init() {
+        add_action('before_woocommerce_init', function() {
+            if (class_exists(FeaturesUtil::class)) {
+                FeaturesUtil::declare_compatibility(
+                    'custom_order_tables', $this->getPluginFile(), true
```
Looks like the third argument is the default value here, so you don't need to specify it.

apeschar reviewed (changes requested):
Looks pretty clean, and I think we're almost there.

to6ko12 committed:
Remove unnecessary parameters
